src/cmdc_tools/datasets/official/CA/data.py

```diff
@@ -1,37 +1,57 @@
-import pandas as pd
+import requests
 import textwrap
 
+import pandas as pd
 import us
 
 from ..base import CountyData
 from ... import DatasetBaseNoDate
 
 
-CA_COUNTY_URL = "https://data.chhs.ca.gov/"
-CA_COUNTY_URL += "dataset/6882c390-b2d7-4b9a-aefa-2068cee63e47/resource/"
-CA_COUNTY_URL += "6cd8d424-dfaa-4bdd-9410-a3d656e1176e/download/covid19data.csv"
-
-
 class CACountyData(DatasetBaseNoDate, CountyData):
     source = "https://public.tableau.com/profile/ca.open.data#!/vizhome/COVID-19HospitalsDashboard/Hospitals"
     state_fips = int(us.states.lookup("California").fips)
     has_fips = False
-    base_url = "https://data.ca.gov/dataset/"
+    query_url = "https://data.ca.gov/api/3/action/datastore_search"
+
+    def data_from_api(self, resource_id, limit=1000, **kwargs):
+        # Create values needed for iterating
+        offset = 0
+        nrecords = 0
+        params = dict(resource_id=resource_id, limit=limit, offset=offset, **kwargs)
+
+        dfs = []
+        keep_requesting = True
+        while keep_requesting:
+            res = requests.get(self.query_url, params=params).json()
+            if not res["success"]:
+                raise ValueError("The request open CA data request failed...")
+
+            records = res["result"]["records"]
+            offset += len(records)
+            keep_requesting = offset < res["result"]["total"]
+
+            dfs.append(pd.DataFrame(records))
+            params.update(dict(offset=offset))
+
+        out = pd.concat(dfs, axis=0, ignore_index=True)
+
+        return out
 
     def get(self):
         cd_df = self.get_case_death_data()
         hosp_df = self.get_hospital_data()
-        df = pd.concat([cd_df, hosp_df], axis=0, ignore_index=True)
+        test_df = self.get_test_data()
+
+        df = pd.concat([cd_df, hosp_df, test_df], axis=0, ignore_index=True, sort=True)
         df["value"] = df["value"].astype(int)
         df["vintage"] = pd.Timestamp.utcnow().normalize()
 
         return df
 
     def get_case_death_data(self):
-        url = self.base_url + "590188d5-8545-4c93-a9a0-e230f0db7290/resource/"
-        url += "926fd08f-cc91-4828-af38-bd45de97f8c3/download/"
-        url += "statewide_cases.csv"
-        df = pd.read_csv(url, parse_dates=["date"])
+        resource_id = "926fd08f-cc91-4828-af38-bd45de97f8c3"
+        df = self.data_from_api(resource_id=resource_id)
```

Comment:
I'll be honest, I  don't totally understand the move to an api vs downloading the file.

Was this broken using the file download?

Reply:
Not broken. I felt the API would likely be more stable than the csv download link, but it wasn't crucial.